File: 20170407/API_NetworkTest/API_NetworkTest/ViewController.m
//
//  ViewController.m
//  API_NetworkTest
//
//  Created by ji jun young on 2017. 4. 7..
//  Copyright © 2017년 Fastcampus. All rights reserved.
//

#import "ViewController.h"
#import "NetworkModule.h"
#import "DataCenter.h"
#import "MainTableViewCell.h"

@interface ViewController ()
<UITableViewDelegate, UITableViewDataSource>

@property (weak, nonatomic) IBOutlet UIButton *locationButton;
@property (weak, nonatomic) IBOutlet UIButton *categoryButton;
@property (weak, nonatomic) IBOutlet UITableView *mainTableView;
@property (weak, nonatomic) UIImageView *headerImageView;


@end

@implementation ViewController


- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.

//    [NetworkModule getPostListWithCompletionBlock:^(BOOL isSuccess, NSDictionary *result) {
//        
//        NSLog(@"viewdidload에서 네트워크 모듈 메소드를 호출함");
//    }];
    
//    [[DataCenter sharedInstance] receivingServerData];
    
//    [[DataCenter sharedInstance] changingArraytoDic];

    
    /**************** tableView Setting ********************************/
    
    //    UITableView *mainTableView = [[UITableView alloc]initWithFrame:CGRectMake(0, 100, self.view.frame.size.width, self.view.frame.size.height-100) style:UITableViewStylePlain];
    //    self.mainTableView = mainTableView;
    self.mainTableView.delegate = self;
    self.mainTableView.dataSource = self;
    //    [self.view addSubview:self.mainTableView];
        [[DataCenter sharedInstance]receiveServerDataWithCompletionBlock:^(BOOL isSuccess) {
            if (isSuccess) {
                dispatch_async(dispatch_get_main_queue(), ^{
                    [self.mainTableView reloadData];
                    NSLog(@"ReceivingServerData and ReloadingData is Completed");
                });
            }else{
                NSLog(@"ReceivingServerData and ReloadingData is Failed");
            }
        }];

     /**************** Deprecated due to if/else issue ***********************/
    //    [[DataCenter sharedInstance] receivingServerDatawithCompletionBlock:^{
    //
    //        dispatch_async(dispatch_get_main_queue(), ^{
    //            [self.mainTableView reloadData];
    //        });
    //
    //        NSLog(@"receivingServerDatawithCompletionBlock, reloadData");
    //    }];

}

/**************** tableviewDelegate ********************************/

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView{
    return 1;
}

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    
    return ([DataCenter sharedInstance].networkDataArray.count);
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    /**************** Deprecated. currentRow is replaced by networkDataArray ***********************/
//    [DataCenter sharedInstance].currentRow = indexPath.row;
    MainTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell"];
    
    if (cell == nil) {
    cell = [[MainTableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"cell"];
    }
    
    /**************** changing cell text with networkDataArray ********************************/
    NSDictionary *temp = [[DataCenter sharedInstance].networkDataArray objectAtIndex:indexPath.row];
    
    //cell이 별도의 메소드를 사용해서 불러오는 것 보다는 하단의 방법으로 텍스트와 이미지를 바꾸는 것이 더 낫다고 함
    //왜냐하면 tableview의 dataSource가 self(ViewController) 이므로
    cell.tutorNameLabel.text = [[temp objectForKey:@"tutor"] objectForKey:@"name"];
    cell.titleLabel.text = [temp objectForKey:@"title"];
    cell.tuteeCountNumberLabel.text = [NSString stringWithFormat:@"%@", [temp objectForKey:@"registration_count"]];
    
    /**************** changing cell image with networkDataArray ********************************/
    NSString *titleImageViewURL = [NSString stringWithFormat:@"%@", [temp objectForKey:@"cover_image"]];
    NSURL *titleURL = [NSURL URLWithString:titleImageViewURL];
    cell.titleImageView.image = [UIImage imageWithData:[NSData dataWithContentsOfURL:titleURL]];
    
    NSString *profileImageViewURL = [NSString stringWithFormat:@"%@", [[temp objectForKey:@"tutor"] objectForKey:@"profile_image"]];
    NSURL *profileURL = [NSURL URLWithString:profileImageViewURL];
    cell.profileImageView.image = [UIImage imageWithData:[NSData dataWithContentsOfURL:profileURL]];
    
    //해당 코드는 메인스레드에서 돌아가는 코드임...따라서 이미지를 여러개 불러올 경우엔 이미지가 다 불러질 때까지 기다리게 됨...따라서 ui도 안먹히게 됨...즉 비동기 처리를 해줘야 함
    
//    NSLog(@"%ld", [DataCenter sharedInstance].currentRow);
    
    /**************** Deprecated due to setting tableView Data issue ***********************/
//    [cell settingText:indexPath];
    cell.selectionStyle = UITableViewCellSelectionStyleNone;
    
    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath{
//    MainTableViewCell *cell= [[MainTableViewCell alloc]init];
//    heightForRow = cell.mainView.frame.size.height;
    
//    MainTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell"];
//    
//    return cell.frame.size.height;

    
    return 200;
}





- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: 20170618/classAndInstance/classAndInstance/ViewController.swift
//
//  ViewController.swift
//  classAndInstance
//
//  Created by ji jun young on 2017. 6. 18..
//  Copyright © 2017년 Fastcampus. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        //데이터 모델의 중요한 구성요소로써 Object를 만든다.
        //클래스는 객체지향 프로그래밍(OOP, Object Oriented Programming)의 바탕을 이룬다.
        //오브젝트는 참조로 작동된다..."주소값만 공유된다..."구조체는 할당될 때 그 시점의 값만 복사된다"
        //
        //구조체로 만든 것은 오브젝트라고 안하고 클래스로 만든 것을 오브젝트라고 하나...
        //스위프트에서는 구조체, 클래스로 만든 것들을 통틀어서 인스턴스라고 한다///
        //
        //클래스로 만든 오브젝트는 Reference Type으로 동작하며, 할당시 복사되지 않는다.
        //하나의 인스턴스에 대한 레퍼런스를 공유한다.
        //인스턴스를 할당한 뒤 수정하면 원본과 할당된 곳 모두 수정됨
        //let으로 정의 해도 프로퍼티가 var이면 수정 가능
        
        struct taskStruct{
            var title : String
            var time : Int?
            var owner : Employee
            var paticipant : Employee?
            
        }
        
        class Employee{
            var name : String?
            var phone : String?
            var boss : Employee?
        }
        
        let me:Employee = Employee()
        me.name = "alex"
        me.phone = "0000"
        
        let toby = Employee()
        //상단의 let me:Employee = Employee() 이든 해당 코드이든, 타입 자체는 어느정도는 알아서 잡아주므로 둘다 인스턴스가 만들어진다
        toby.name = "toby"
        toby.phone = "1111"
        print(toby.phone!)
        
        //me와 toby가 let인 이유?
        //구조체의 인스턴스는 let으로 만들면 아무것도 수정할 수가 없다
        //그러나 클래스의 인스턴스는 let으로 만들어도 나중에 수정을 할 수 있다 = let으로 정의 해도 프로퍼티가 var이면 수정 가능
        
        var callTask = taskStruct(title : "call to toby", time:90, owner:me, paticipant:toby)
        var reportTask = taskStruct(title : "call to boss", time:nil, owner:me, paticipant:nil)
        
        
        callTask.paticipant?.phone = "2222"
        print(toby.phone!)
        //particiant인 toby는 클래스 이므로, 구조체로 접근해서 phone을 수정했더라도 클래스 안의 프로퍼티인 phone도 같이 수정이 되어 버림
        
        
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


}

